src/Parser.php

```diff
@@ -363,18 +358,37 @@ protected function isIpv6Host(string $ipv6): bool
      */
     protected function isRegisteredName(string $host): bool
     {
-        if ('.' === substr($host, -1, 1)) {
-            $host = substr($host, 0, -1);
+        // Note that unreserved is purposely missing . and -
+        // . is used to separate labels
+        // - cannot start or end a label
+        static $pattern = '/(?(DEFINE)
+                (?<unreserved> [a-z0-9\_\~])
+                (?<gen_delims> [:\/?#\[\]@])
+                (?<sub_delims> [!$&\'()*+,;=])
+                (?<label> (?:(?&unreserved)(?:(?&unreserved)|(?&gen_delims)|(?&sub_delims)|\-){0,61})?(?&unreserved))
+            )
+            ^(?:(?&label)\.){0,126}(?&label)\.?$/imx';
+
+        if (preg_match($pattern, $host)) {
+            return true;
         }
 
-        $labels = array_map([$this, 'toAscii'], explode('.', $host));
+        if (false !== strpos($host, '%')) {
+            $host = rawurldecode($host);
+        }
 
-        return 127 > count($labels)
-            && 253 > strlen(implode('.', $labels))
-            && $labels === array_filter($labels, [$this, 'isHostLabel']);
+        $host = idn_to_ascii($host, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
```

Comment:
It seems like IDNs are not covered by the defined `reg-name` syntax in https://tools.ietf.org/html/rfc3986#section-3.2.2

>     reg-name    = *( unreserved / pct-encoded / sub-delims )

This means these hosts are strictly coupled to DNS, which means we can apply all DNS restrictions on IDN domains, but not on other host names.

Reply:
IDN domains are defined in rfc3987 so yes they do not cover all the reg name they only cover i18n cases